### ISSUE TITLE
Feature request: unload env vars after use?

### ISSUE THREAD

**Issue by starkeepers** [fireproofsocks/dotenvy#4]
Tentative feature request: for additional security, would you consider (optionally) unloading the loaded env vars after startup? That would ensure that sensitive data doesn't remain in the environment and hence cannot be dumped/read later on.

**Comment by fireproofsocks** [comment]
Can you include steps to reproduce exactly what you're talking about?  When the side-effect of `System.put_env/1` is called, that only affects the current OS process, so they should not remain once the process completes (i.e. once the app is terminated).  You can override the `side_effect` option for `Dotenvy/source/2` so that the values are not put into the system ENV, and instead you could work with the returned values, e.g. something like:

```
{:ok, vals} = Dotenvy.source([".env", ".env2", ".env3"], side_effect: false)

config :myapp, :foo, vals["FOO"]
```
Some dependencies may be coded to specifically read system ENVs, however, so your mileage may vary.

**Comment by starkeepers** [comment]
> that only affects the current OS process, so they should not remain once the process completes (i.e. once the app is terminated)

1. Since elixir/phoenix apps tend to be long running, this means there is now a long-running OS process with env vars containing sensitive/confidential data. Which is a potential vulnerability. 

2. that data is typically only needed at app-startup time, so an apparent solution would be for the start function to unload/unset the env vars after it has read them

3. to maximise the likelihood of this being done, and done correctly, the ideal case would be for dotenvy to offer it as an option, something like `unload_after_start: true` in the config file.

I hope this clarifies it, and again, just an idea!

**Comment by fireproofsocks** [comment]
Does the workaround work for you?  I.e. if you bypass the side-effect, then the variables are never put into the system env.  
Alternatively, you could probably flush out the system envs at the bottom of your `runtime.exs` (untested), but I'd still like to see a viable way to view the envs outside the running process, i.e. in places where you'd have to have access to the running app instance and `Application.get_env`.  In other words, I'm not entirely sure there's a security win to be had by flushing the system env when the application config is still accessible.  I'm happy harden things if appropriate, however.

**Comment by starkeepers** [comment]
fwiw the emphasis was on point 3, ie making it easy and maybe even default to take the more secure path, rather than each dev hand-coding a workaround (but thanks for pointing out the workarounds -- they do work for me). 

The kind of issues I had in mind are summarised in eg: [Why you shouldn't use ENV variables for secret data](https://diogomonica.com/2017/03/27/why-you-shouldnt-use-env-variables-for-secret-data/)

ps. the `side_effect: false` solution comes very close to what I envisaged, so maybe this request mostly comes down to a request to make this solution more prominent in the documentation/readme?

**Comment by fireproofsocks** [comment]
That all makes sense, and I think it makes sense to leave the feature request open for now -- security is a big topic and thank you for highlighting it.  I'm not sure how to best implement a solution, however.  So far, every scenario I can think of where I'd be able to read some value happens in a situation where I have privileged access anyway (i.e. where I'd have more to worry about than just an ENV being leaked).  E.g. what good is it to wipe the system envs after app starting if you can't read them unless you're on the box, and if you're on the box and you can remote into the app, you can still do all the things that make use of those values (including printing them)?  It's totally possible there are angles to this that I'm not considering, however.

I'm more of the mind that the best way to "keep" secrets is not to have them.  In practical terms, that usually means that devs provide their own "throw-away" values (e.g. sandbox API tokens or SSH keys with limited permissions) and that the management of the _real_ sensitive values during the deployment is paramount (i.e. when the _live_ API tokens or SSH keys are used, that process is tightly managed). Although it feels like this approach side-steps the issue, I think it's actually the correct approach because it shifts the onus of safeguarding the sensitive info to the deployment process because ultimately, that's where things must be locked down.  E.g. implementation of encryption (e.g. via https://github.com/kraftjectory/secrex) requires that _somebody_ safeguards the key file.

**Comment by starkeepers** [comment]
I agree that "the best way to "keep" secrets is not to have them" :) That doesn't mean that we shouldn't help in the case where people do have them, and I think that the line of thought of "I have privileged access anyway" misses the point that security is a matter of gradient, not absolutes, and hence any additional protection is helpful. For instance, the article I posted mentions that some of the issues come from unintentional leakage of env vars, and those are prevented by the env vars being automatically unset after being read (or never being set, as in `side_effect: false`).

That said, it indeed seems wise to wait until you feel comfortable with an implementation (if at all). And thanks for the reference to secrex, I didn't know about it.

**Comment by lud** [comment]
You could add a new function `local_source(files)` that would write the env vars to the process dictionnary (with respect to the overwrite configuration of course, i.e. do not write a variable if already present in env), and the `env!` function would look here before calling `System.fetch_env`.

The process executing `runtime.exs` exits quickly after the config is set.

**Comment by fireproofsocks** [comment]
I keep thinking about how to best implement this.  Possibly the easiest first step would be to implement a function like @lud described: one whose side effect is to write to the `Application` process dictionary (when I first sketched out this package, I had included a `merge/2` function that would simply merge values _without_ a side-effect, so this idea isn't so foreign).  Per @starkeepers points, offering a function like that would help avoiding unintended (i.e. insecure) side effects.

I'm not as clear about adding multiple data sources to the `env!/2` function, however.  When  a value might come from multiple places it becomes less what the function is actually doing -- you might end up with a headache debugging where values are actually coming from.  The source code for the `env!/2` function could easily be modified if the `System` module were replaced with the `Application` module (thus `Application.fetch_env` instead of `System.fetch_env` etc).  I'm wondering if it might be acceptable UI/UX if you had to do `import Dotenvy.Secure` or maybe even tap into `use` by doing something like `use Dotenvy, store: Application`.  There are a few other considerations that might pop up there (specifically because System ENV always arrive as strings, whereas items stored in the process dictionary may be any valid Elixir term), but I thought it would be worthwhile mentioning the idea here in the relevant ticket.

**Comment by lud** [comment]
To keep the library simple and master at one single feature, I think it would be best to keep the separation between the two main points: read .env files that provide strings, and cast string values to data types. And let users add their logic with that.

Another thing to consider is that Dotenvy is not responsible for variables that are already in the environment when the config is loaded. If such variables were to be deleted, the developer can simply use System.detele_env.

To me, Dotenvy is only responsible for its side effects, i.e. creating and/or overwriting environment variables.

We could imagine a simple system where the side effect function would return either `:ok` or a `cleanup` function that should reverse the side effect. The developers could then provide their own side effect and cleanup function. The default implementation for dotenvy would consist on keeping all variables names that were created, and variables overwritten names and their previous value, and return a fun that would delete the added variables and put the old values back for the overwritten ones.

It would be up to the user of the library, the developers, to call the cleanup function or not.

Problem with that is that you need either a different function to get the variables as a map (typical usage with side_effect==false), or you need to return `{:ok, vars, cleanup}`

**Comment by starkeepers** [comment]
> the easiest first step would be to implement a function like @lud described: 
> one whose side effect is to write to the Application process dictionary 
> (when I first sketched out this package, I had included a merge/2 function 
> that would simply merge values without a side-effect, so this idea isn't so foreign). 
> Per @starkeepers points, offering a function like that would help avoiding unintended (i.e. insecure) side effects.

That sounds great. Maybe there could be variant of the same name with !, ie `Dotenvy.source` and `Dotenvy.source!`, where the former is "safe" while the latter is "unsafe" (on analogy with "can raise errors")? Or vice-versa, the bang version could bring attention to the fact that it writes to the process dictionary.

(To avoid the "reading from multiple sources" problem, is it an option to write the merged preexisting env + newly-read env into the local process dictionary, and then have that as the unique source of truth for env! ? )

**Comment by lud** [comment]
>>> (To avoid the "reading from multiple sources" problem, is it an option to write the merged preexisting env + newly-read env into the local process dictionary, and then have that as the unique source of truth for env! ? )

This would make impossible to read those values from other processes.

**Comment by fireproofsocks** [comment]
Thank you @starkeepers and @lud for the continued discussion; I appreciate your input and feedback.

Is there any concern about writing values to the Application's process dictionary?  It is redundant storage because the values will be eventually read into other locations within the same dictionary, so should we consider the effects of loading up large numbers of variables?  I'm unclear on whether storing the variables in the process dictionary offers any greater security than storing them in the OS process' system ENVs.  Regardless, I don't see any difficulty in refactoring the existing `source` functions to put values into the Application process dictionary and adapting the `env!` functions to read from it.  

That might obviate the need for a `merge` function entirely.

**Comment by fireproofsocks** [comment]
The more I think about the `merge` function, the more I think that it might be subtly smelly.  E.g. if you were handed a merged bunch of values, they would all be strings, so you'd need to convert them.  You could either do something like 
```
{:ok, vars} = merge([".env1", ".env2", ...])

config :myapp, 
  :host, Dotenvy.Transformer.to!(vars["HOST"], :string),
  :port, Dotenvy.Transformer.to!(vars["PORT"], :integer!)
```
OR, you could transform merged values via a `transforms` option, e.g.

```
{:ok, vars} = merge([".env1", ".env2", ...], transforms: %{
    "HOST" => :string, 
    "PORT" => :integer?, 
    "OPTIONAL" => [:module, SomeDefaultValue]
})

config :myapp, 
  :host, Map.fetch!(vars, "HOST")
```
but this also feels inelegant.

Let's start with the changes that will write vars to the process dictionary and go from there.

**Comment by dch** [comment]
> for additional security, would you consider (optionally) unloading the loaded env vars after startup?

Unfortunately the original environment variables are available to all users with appropriate file system permissions in `/proc/:pid/environ` on Linux,

> That would ensure that sensitive data doesn't remain in the environment and hence cannot be dumped/read later on.

Any user with elevated privileges can trivially read the memory space of any process and look for discarded secrets. The BEAM doesn't provide a way to zero out a previous variable, or selected block of memory, to erase the secrets.

See https://elixirforum.com/t/keeping-system-envs-secure/41167 for a somewhat longer reply, including some possible alternatives.

**Comment by starkeepers** [comment]
> Let's start with the changes that will write vars to the process dictionary and go from there.

That sounds great. 

As for the rest, I'm not sure we're all taking about the same thing, it seems to me that runtime.exs could look almost identical to before, ie:

```
import Config
import Dotenvy

# 1. read the .env files, merge them with the env vars from the current unix process, 
# and store the results in an elixir-process-private way
# (which can be a private ets table, the process dictionary, whatever):

read_env([".env", ".env2", ".env3"])

# 2. use those values as usual, with env! now modified to read from the elixir-process-private data:

config :myapp, MyApp.Repo,
    database: env!("DATABASE", :string!),
    pool_size: env!("POOL_SIZE", :integer),
    password: env!("PASSWORD", :string)

# 3. at the end of runtime.exs, the associated process dies, so the values read
# from .env/.env2/.env3 vanish with it, leaving no insecure trace,
# not accessible from any other process or surviving anywhere, which was the point.
```

Any other way of doing it is fine too, if it facilitates not having sensitive values in the unix process environment after use (thereby eliminating at least one low-hanging-fruit subclass of security issues).

ps. I've used the `read_env` terminology here instead of `source!` or `local_source` since `source` has such a strongly associated semantics of loading values into the environment. Maybe the `read(_env)` notion is more descriptive of what this would be doing, namely only reading the values, without writing them into the environment.

**Comment by starkeepers** [comment]
> The BEAM doesn't provide a way to zero out a previous variable

Thanks! So another reason to never put  the content of the files sourced by dotenvy into the unix-process environment in the first place, and instead keep it in ets, a process-local dictionary, a gen-server, or whatever other mechanism for local-state retention.

**Comment by fireproofsocks** [comment]
Thank you @dch , @starkeepers , and @lud for your input.

I have a PR up that implements changes in the side_effect: https://github.com/fireproofsocks/dotenvy/pull/8
If anyone has time/interest to check it out and provide feedback that would be great -- I'll leave that PR open for a day or so before I merge and tag it just in case additional changes are required.

Thanks again!

**Comment by lud** [comment]
Application.put_env makes the data available globally, it is not using the process dictionary of the calling process. Plus now you have to call source even if there is no .env file, because Dotenvy does not read from system environment variables.

**Comment by fireproofsocks** [comment]
Ha!  Good point -- I sheepishly admit that I had conflated the `Application` with the `Process` dictionary.  One more commit pushed up...

**Comment by fireproofsocks** [comment]
I've released v0.5.0 with PR #8 
Let's open a new issue if further refinements/improvements/fixes are needed.
